packages/financial-templates-lib/src/price-feed/CoinGeckoPriceFeed.js

```diff
@@ -0,0 +1,158 @@
+    this.invertPrice = invertPrice;
+    this.priceFeedDecimals = priceFeedDecimals;
+
+    this.priceHistory = []; // array of { time: number, price: BN }
+  }
+
+  async update() {
+    const currentTime = this.getTime();
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== undefined && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
+      this.logger.debug({
+        at: "CoinGeckoPriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTimes + this.minTimeBetweenUpdates - currentTime
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "CoinGeckoPriceFeed",
+      message: "Updating CoinGeckoPriceFeed",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime
+    });
+
+    // 1. Construct URL.
+    // See https://www.coingecko.com/api/documentations/v3#/operations/simple/get_simple_token_price__id_ for how this url is constructed.
+    const url =
+      "https://api.coingecko.com/api/v3/simple/token_price/ethereum" +
+      `?contract_addresses=${this.contractAddress}&vs_currencies=${this.currency}`;
+
+    // 2. Send request.
+    const response = await this.networker.getJson(url);
+
+    // 3. Check response.
+    if (!response || !response[this.contractAddress] || !response[this.contractAddress][this.currency]) {
+      throw new Error(`🚨Could not parse result from url ${url}: ${JSON.stringify(response)}`);
+    }
+
+    // 4. Parse result.
+    // Return data structure:
+    // {
+    //   "<contractAddress>": {
+    //     "<currency>": <price>
+    //   }
+    // }
+    const newPrice = this._convertPriceFeedDecimals(response[this.contractAddress][this.currency]);
+
+    // 5. Store results.
+    this.currentPrice = newPrice;
+    this.lastUpdateTime = currentTime;
+    this.priceHistory.push({ time: currentTime, price: newPrice });
+  }
+
+  getCurrentPrice() {
+    return this.invertPrice ? this._invertPriceSafely(this.currentPrice) : this.currentPrice;
+  }
+
+  getHistoricalPrice(time) {
+    if (this.lastUpdateTime === undefined) {
+      return undefined;
+    }
+
+    let matchingPrice;
+    for (const history of this.priceHistory) {
+      const minTime = history.time - this.lookback;
+      const maxTime = history.time;
+
+      if (time >= minTime && time <= maxTime) {
+        matchingPrice = history.price;
+        break;
+      }
+    }
+
+    if (!matchingPrice) {
+      return undefined;
+    }
+
+    return this.invertPrice ? this._invertPriceSafely(matchingPrice) : matchingPrice;
+  }
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  getPriceFeedDecimals() {
+    return this.priceFeedDecimals;
+  }
+
+  getLookback() {
+    return this.lookback;
+  }
+
+  _convertPriceFeedDecimals(number) {
```

Comment:
to follow the pattern done in other price feeds, please define this method at the top within your constructor as `this.convertPriceFeedDecimals`. see: https://github.com/UMAprotocol/protocol/blob/9fcc5128fe3d684f4a87e2efa3d2e49934b96766/packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.js#L56

Reply:
@chrismaree sure we can follow this convention. But out of curiosity, is there any particular reason why the function is defined in the constructor instead?

Reply:
Fixed :)